tower-http/src/services/fs/read_stream.rs

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright (c) 2021 Tokio Contributors
+
+Permission is hereby granted, free of charge, to any
+person obtaining a copy of this software and associated
+documentation files (the "Software"), to deal in the
+Software without restriction, including without
+limitation the rights to use, copy, modify, merge,
+publish, distribute, sublicense, and/or sell copies of
+the Software, and to permit persons to whom the Software
+is furnished to do so, subject to the following
+conditions:
+
+The above copyright notice and this permission notice
+shall be included in all copies or substantial portions
+of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+ */
+use bytes::{Bytes, BytesMut};
+use futures_core::stream::Stream;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use pin_project_lite::pin_project;
+use tokio::io::AsyncRead;
+use tokio_util::io::poll_read_buf;
+
+const DEFAULT_CAPACITY: usize = 4096;
+
+pin_project! {
+    #[derive(Debug)]
+    pub(crate) struct ReaderStream<R> {
+        // Reader itself.
+        //
+        // This value is `None` if the stream has terminated.
+        #[pin]
+        reader: Option<R>,
+        // Working buffer, used to optimize allocations.
+        buf: BytesMut,
+        capacity: usize,
+        max_read_bytes: Option<usize>,
+        read_bytes: usize,
+    }
+}
```

Comment:
I've just noticed that `AsyncReadExt` has a [`take`](https://docs.rs/tokio/1.14.0/tokio/io/trait.AsyncReadExt.html#method.take) adapter.

> Creates an adaptor which reads at most `limit` bytes from it.

Do you think we could use that instead of this custom version of `ReaderStream`?